### PR TITLE
fix: set ownerReference on imageRegistry and globalMirror Secrets

### PR DESCRIPTION
**Pull request by dkoshkin**
**What problem does this PR solve?**:
Add `ownerReference` to user provided credentials Secrets for `imageRegistry` and `globalMirror` variables. These Secrets will now be moved with `clusterctl move` and deleted once the Cluster is deleted.

I needed `EnsureOwnerRefForSecret ` from the lifecycleutils package but that function needed alot of other functions from that package, so moved it all up the handlers package. (first commit)

**Which issue(s) this PR fixes**:
Fixes https://jira.nutanix.com/browse/NCN-100641

**How Has This Been Tested?**:

Added unit tests and tested manually for both `imageRegistry` and `globalMirror`:
```
$ kubectl get secret dkoshkin-registry-creds-owner-ref-1-dockerhub-credentials -o yaml
apiVersion: v1
data:
  password: <>
  username: <>
kind: Secret
metadata:
  creationTimestamp: "2024-06-13T23:47:59Z"
  labels:
    cluster.x-k8s.io/provider: nutanix
  name: dkoshkin-registry-creds-owner-ref-1-dockerhub-credentials
  namespace: default
  ownerReferences:
  - apiVersion: cluster.x-k8s.io/v1beta1
    kind: Cluster
    name: dkoshkin-registry-creds-owner-ref-1
    uid: be432689-315c-4511-a294-f397b3cdb9b7
  resourceVersion: "3886"
  uid: 84366a22-63db-4820-bf31-b5bf36e2c6cb
type: Opaque
```

**Special notes for your reviewer**:

### CODE REVIEW

File: common/pkg/testutils/capitest/patches.go [hunk 1]
```diff
@@ -29,6 +29,7 @@ type PatchTestDef struct {
 	ExpectedPatchMatchers []JSONPatchMatcher
 	// UnexpectedPatchMatchers used to test patches that should not be present
 	UnexpectedPatchMatchers []JSONPatchMatcher
+	AdditionalValidations   func()
```
Comment (dkoshkin) [inline common/pkg/testutils/capitest/patches.go:32]: Added this in `PatchTestDef` as a generic function to validate anything else for these handler tests, but let me know if we don't like this.

File: pkg/handlers/generic/mutation/imageregistries/credentials/inject_test.go [hunk 2]
```diff
@@ -140,254 +142,271 @@ var _ = Describe("Generate Image registry patches", func() {
...
 	}
 
-	testDefs := []capitest.PatchTestDef{
+	testDefs := []struct {
+		capitest.PatchTestDef
+		assertOwnerReference bool
+	}{
 		{
-			Name: "unset variable",
+			PatchTestDef: capitest.PatchTestDef{
```
Comment (dkoshkin) [inline pkg/handlers/generic/mutation/imageregistries/credentials/inject_test.go:150]: No real change to the input here, just needed to indent existing tests under `capitest.PatchTestDef`

**Review by jimmidyson — APPROVED** [review]

File: pkg/handlers/generic/mutation/imageregistries/credentials/inject_test.go [hunk 3]
```diff
@@ -436,7 +455,21 @@ var _ = Describe("Generate Image registry patches", func() {
...
+				client, err := helpers.TestEnv.GetK8sClientWithScheme(clientScheme)
+				gomega.Expect(err).To(gomega.BeNil())
+
+				secret := newRegistryCredentialsSecret(validSecretName, request.Namespace)
+				gomega.Expect(client.Get(
+					context.Background(),
+					ctrlclient.ObjectKeyFromObject(secret),
+					secret,
+				)).To(gomega.BeNil())
+				gomega.Expect(secret.OwnerReferences).ToNot(gomega.BeEmpty())
```
Comment (dlipovetsky) [inline pkg/handlers/generic/mutation/imageregistries/credentials/inject_test.go]: Did you consider asserting the owner kind (and maybe name)?

**Review by dlipovetsky — APPROVED** [review]

File: pkg/handlers/generic/mutation/imageregistries/credentials/inject_test.go (hunk 3 above)
Comment (dkoshkin) [reply pkg/handlers/generic/mutation/imageregistries/credentials/inject_test.go]: Good idea, added in https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/720/commits/d7920c985fb8c764cb7e43bc26085993ee437cfd

File: pkg/handlers/generic/mutation/imageregistries/credentials/inject.go [hunk 4]
```diff
@@ -131,6 +158,23 @@ func (h *imageRegistriesPatchHandler) Mutate(
 			registriesWithOptionalCredentials,
 			mirrorCredentials,
 		)
+
+		if secretName := secretNameForImageRegistryCredentials(globalMirror.Credentials); secretName != "" {
```
[the remark this answers is not in the document]
Comment (supershal) [reply pkg/handlers/generic/mutation/imageregistries/credentials/inject.go:162]: same for other calls above made to the api server

[the remark this answers is not in the document]
Comment (dkoshkin) [reply pkg/handlers/generic/mutation/imageregistries/credentials/inject.go]: Yep good call, done in https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/720/commits/fc26132c04d547538581598387974d5eb58dd3de

**Review by supershal — APPROVED** [review]
Thanks for the fix and test.

**Review by supershal — APPROVED** [review]

**Pull request merged**